### ISSUE TITLE
How to accumulate model outputs for across batch losses?

### ISSUE THREAD

**Issue by Tarmily-Wen-Percepta** [PyTorchLightning/pytorch-lightning#2693]
I was hoping that setting accumulate_grad_batches=2 and implementing training_step_end would result in training_step_end receive output from both training_steps, so list of len 2 in args. But I only received the batch output from the latest batch.

The reason I want both batches is because I am using a contrastive loss, and I want to constrast the embedding across and within batches depending on their classes. So essentially if I have 2 batches of size 8, I would like to combine the outputs from 2 training_step to get an effective batch of 16 to perform this contrastive loss.

How could I implement this without needing to manually keep track of the current batch_idx before clearing out some number of stored training_step_outputs

This is my current hack which isn't great since it could be wrong. It leads to potential model updates before computing the contrastive loss making the embeddings no longer valid for the loss.
``` python

if (batch_idx + 1) % self.config.PYTORCH_LIGHTNING.accumulate_grad_batches == 0:
    loss = self.contrastive_loss(self.accumulated_batches)
    self.accumulated_batches = []
else:
    self.accumulated_batches.append(<training_step output>)

```

**Comment by Maddy12** [comment]
Is this going to be addressed in another issue?
